### DISCUSSION TITLE
Call JavaScript from R directly

### DISCUSSION THREAD

**Discussion by shahreyar-abeer** [Appsilon/rhino#323] [Q&A] [answered]
The documentation says we can call JavaScript like this.

### `app/js/index.js`
```js
export function sayHello() { console.log('Hello!'); }
```

### `code.R`
```r
tags$button(onclick = "App.sayHello()")
```

This works nicely.
But I want to call it directly from an `observeEvent` like this

```r
observeEvent(input$button, {
    # run the JavaScript function sayHello
})
```
How do I do that?
I'm looking for something in the line of `session$sendCustomMessage` and not `shinyjs`

**Comment by kamilzyla** [comment, marked as answer]
Hi @shahreyar-abeer, thanks for your question! I'd suggest a solution along these lines:

### `app/js/index.js`
```js
export function sayHello(name) {
  console.log(`Hello ${name}!`);
}

Shiny.addCustomMessageHandler('sayHello', sayHello);
```

### `app/main.R`
```r
box::use(
  shiny,
)

say_hello <- function(name, session = shiny$getDefaultReactiveDomain()) {
  session$sendCustomMessage("sayHello", name)
}

#' @export
ui <- function(id) {
  ns <- shiny$NS(id)
  shiny$bootstrapPage(
    shiny$textInput(ns("name"), "Your Name"),
    shiny$actionButton(ns("hello"), "Say Hello!")
  )
}

#' @export
server <- function(id) {
  shiny$moduleServer(id, function(input, output, session) {
    shiny$observeEvent(input$hello, say_hello(input$name))
  })
}
```

### Notes
The call to `Shiny.addCustomMessageHandler()` in JS and the `sayHello` wrapper in R are a kind of boilerplate code which will be very similar for any such handler you define. We are considering to add a mechanism to Rhino so that such definitions would be included automatically :slightly_smiling_face:

**Reply by shahreyar-abeer** [reply]
Hi @kamilzyla 

Thanks for the answer!
So it's pretty much the same as before.
Just that we define the JS function separately with `export` instead of defining it inside `Shiny.addCustomMessageHandler`

Thanks a lot for the great package!

**Reply by kamilzyla** [reply]
I may add that the `export` keyword can be actually omitted in this case. It's only needed if you want to refer to the function via `App.sayHello` e.g. in an `onclick` handler :wink:

**Reply by shahreyar-abeer** [reply]
Uh I see.
Thanks for the clarification!
